File: languages/python/assessments/02-core/main.py
from pathlib import Path


def build_frequency(scores: list[int]) -> list[tuple[str, int]]:
    buckets = []
    for start in range(0, 100, 10):
        end = 100 if start == 90 else start + 9
        label = f"{start}-{end}"
        count = sum(1 for score in scores if start <= score <= end)
        buckets.append((label, count))
    return buckets


def main() -> None:
    scores = []

    while True:
        raw = input("Enter a score (0-100) or -1 to finish: ").strip()
        try:
            value = int(raw)
        except ValueError:
            print("Invalid integer. Try again.")
            continue

        if value == -1:
            break

        if not 0 <= value <= 100:
            print("Score ignored because it is outside 0-100.")
            continue

        scores.append(value)

    frequency = build_frequency(scores)
    lines = [f"Valid scores: {len(scores)}"]

    if scores:
        average = sum(scores) / len(scores)
        lines.append(f"Average: {average:.2f}")
        lines.append(f"Minimum: {min(scores)}")
        lines.append(f"Maximum: {max(scores)}")
    else:
        lines.append("Average: N/A")
        lines.append("Minimum: N/A")
        lines.append("Maximum: N/A")

    lines.append("Frequency:")
    for label, count in frequency:
        lines.append(f"- {label}: {count}")

    for line in lines:
        print(line)

    report_path = Path("core_assessment_report.txt")
    report_path.write_text("\n".join(lines) + "\n", encoding="utf-8")
    print("Report written to core_assessment_report.txt")


if __name__ == "__main__":
    main()

File: languages/java/assessments/02-core/Main.java
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.ArrayList;
import java.util.List;
import java.util.Locale;
import java.util.Scanner;

public class Main {
    static String bucketLabel(int start) {
        int end = start == 90 ? 100 : start + 9;
        return start + "-" + end;
    }

    static int bucketCount(List<Integer> scores, int start) {
        int end = start == 90 ? 100 : start + 9;
        int count = 0;
        for (int score : scores) {
            if (score >= start && score <= end) {
                count++;
            }
        }
        return count;
    }

    public static void main(String[] args) throws IOException {
        Locale.setDefault(Locale.US);
        Scanner scanner = new Scanner(System.in);
        List<Integer> scores = new ArrayList<>();

        while (true) {
            System.out.print("Enter a score (0-100) or -1 to finish: ");
            if (!scanner.hasNext()) {
                break;
            }

            String raw = scanner.next().trim();
            int value;
            try {
                value = Integer.parseInt(raw);
            } catch (NumberFormatException error) {
                System.out.println("Invalid integer. Try again.");
                continue;
            }

            if (value == -1) {
                break;
            }
            if (value < 0 || value > 100) {
                System.out.println("Score ignored because it is outside 0-100.");
                continue;
            }
            scores.add(value);
        }

        List<String> lines = new ArrayList<>();
        lines.add("Valid scores: " + scores.size());

        if (scores.isEmpty()) {
            lines.add("Average: N/A");
            lines.add("Minimum: N/A");
            lines.add("Maximum: N/A");
        } else {
            int total = 0;
            int minimum = scores.get(0);
            int maximum = scores.get(0);
            for (int score : scores) {
                total += score;
                minimum = Math.min(minimum, score);
                maximum = Math.max(maximum, score);
            }
            lines.add(String.format("Average: %.2f", (double) total / scores.size()));
            lines.add("Minimum: " + minimum);
            lines.add("Maximum: " + maximum);
        }

        lines.add("Frequency:");
        for (int start = 0; start < 100; start += 10) {
            lines.add("- " + bucketLabel(start) + ": " + bucketCount(scores, start));
        }

        for (String line : lines) {
            System.out.println(line);
        }

        Path reportPath = Path.of("core_assessment_report.txt");
        Files.write(reportPath, lines);
        System.out.println("Report written to core_assessment_report.txt");
    }
}
